### PR TITLE
✨ [amp-story-embed] Adds messaging, implements build/layoutcallbacks

### PR DESCRIPTION
**Pull request by Enriqe**
Partial for #24539 

This PR adds:
* Messaging that is needed for communication between the viewer and the STAMPs.

* Adds an intersection observer that calls `layoutCallback()` with a fallback that uses a scroll listener.

Things I'm not sure about:
* Since I'm installing a new package (`@ampproject/viewer-messaging`), yarn.lock has been updated, but I don't know if that should be in my local gitignore or if it should actually be checked in.

Issue tracker #26308

### CODE REVIEW

File: src/amp-story-embed.js [hunk 1]
```diff
@@ -31,6 +33,58 @@ const CSS = `
...
+
+/**
+ * Gets encoded url for viewer usage.
+ * @param {string} href
+ * @return {!Location}
+ */
+function getEncodedUrl(href) {
+  const params = {
+    visibilityState: 'inactive',
+    origin: parseUrlDeprecated(self.location.href).origin,
```
Comment (newmuis) [inline src/amp-story-embed.js]: If `parseUrl` is deprecated, what's the way we're supposed to do this?

File: src/amp-story-embed.js [hunk 2]
```diff
@@ -31,6 +33,58 @@ const CSS = `
...
+  };
+  log('Params: ' + JSON.stringify(params));
```
Comment (newmuis) [inline src/amp-story-embed.js]: Is this for debugging?  Should it be removed before merging?

File: src/amp-story-embed.js [hunk 3]
```diff
@@ -31,6 +33,58 @@ const CSS = `
...
+  }
+  return parseUrlDeprecated(inputUrl);
+}
+
+/**
+ * Encodes string of URL parameters.
+ * @param {!Object} params
+ * @return {string}
+ */
+function encodeUrl(params) {
```
Comment (newmuis) [inline src/amp-story-embed.js]: Can't you use this util instead of writing your own?

https://github.com/ampproject/amphtml/blob/2d23dd4bedd6c0bd1af8664ca82c8120a2595247/src/url.js#L262-L279

File: src/amp-story-embed.js [hunk 4]
```diff
@@ -102,6 +159,37 @@ export class AmpStoryEmbed {
...
+      frameOrigin
+    ).then(
+      messaging => {
+        // TODO(Enriqe): Appropiately set visibility to stories.
+        messaging.sendRequest('visibilitychange', {state: 'visible'}, true);
+        messaging.setDefaultHandler(() => {
+          // TODO(Enriqe): Set default handler.
+        });
+        messaging.registerHandler('moreInfoLinkUrl', () => {
+          // TODO(Enriqe): Get publisher defined `moreInfoLinkUrl` (could come
```
Comment (newmuis) [inline src/amp-story-embed.js]: I don't think we actively need a TODO for this; we could just wait until we've gotten a feature request for this.

File: src/amp-story-embed.js [hunk 5]
```diff
@@ -123,23 +211,32 @@ export class AmpStoryEmbed {
...
    */
-  layoutIframe_(index) {
-    const story = this.stories_[index];
-    this.iframeEl_.setAttribute('src', story.href);
-    return Promise.resolve();
+  layoutIframe_(story, iframe) {
+    this.stories_ = toArray(this.element_.querySelectorAll('a'));
+
+    const url = getEncodedUrl(story.href).href;
+    log('AMP URL = ', url);
```
Comment (newmuis) [inline src/amp-story-embed.js]: Should this be removed?

File: src/amp-story-embed.js [hunk 6]
```diff
@@ -150,7 +247,18 @@ self.onload = () => {
...
+    const intersectingCallback = entries => {
+      entries.forEach(entry => {
+        if (!entry.isIntersecting) {
+          return;
+        }
+        embedImpl.layoutCallback();
+      });
+    };
+    const observer = new IntersectionObserver(intersectingCallback, {
+      threshold: 0.5,
```
Comment (newmuis) [inline src/amp-story-embed.js]: I think we might want to consider lowering this threshold considerably.  In fact, I think we probably want to preload `n` viewports in advance.

File: src/amp-story-embed.js (hunk 4 above)
Comment (gmajoulet) [reply src/amp-story-embed.js]: +1, this whole block (L178-L186) is not needed in this PR. Let's introduce these methods properly with their own PR, `visibilitychange` or `moreInforLinkUrl` should not live in the `initializeHandshake_()` method

File: src/amp-story-embed.js [hunk 7]
```diff
@@ -123,23 +211,32 @@ export class AmpStoryEmbed {
...
-   * @return {!Promise}
+   * @param {!Element} story
+   * @param {!Element} iframe
    * @private
    */
-  layoutIframe_(index) {
-    const story = this.stories_[index];
-    this.iframeEl_.setAttribute('src', story.href);
-    return Promise.resolve();
+  layoutIframe_(story, iframe) {
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: The expected behavior for this library is:
1. Load one story, prerender the next story, do nothing with all the other ones
2. When navigating to the next story, prerender the following one

This PR/method is setting you on the wrong track, having a mechanism that loads one story is closer to the end goal than a mechanism that loads all the stories, as you can re-use the code to load the n+1 story.

File: src/amp-story-embed.js (hunk 6 above)
Comment (gmajoulet) [reply src/amp-story-embed.js]: +1, we should do 2 viewports to match the AMP behavior. We should also have a fallback for when `IntersectionObserver` isn't supported.

File: src/amp-story-embed.js (hunk 1 above)
Comment (choumx) [reply src/amp-story-embed.js]: `Services.urlForDoc` has the same function. 

@jridgewell Any end in sight for this migration? :)

File: package.json [hunk 8]
```diff
@@ -20,6 +20,7 @@
   "dependencies": {
     "@ampproject/animations": "0.2.0",
     "@ampproject/worker-dom": "0.22.0",
+    "@ampproject/viewer-messaging": "^1.1.0",
```
[the remark this answers is not in the document]
Comment (choumx) [reply package.json]: We should always use fixed versions (Renovate is also configured to use fixed versions). I think "no ^" might be a lint check?

File: src/amp-story-embed.js (hunk 1 above)
Comment (newmuis) [reply src/amp-story-embed.js]: Ah, we can't use services here, I think, since v0 won't be on the page.

Comment (Enriqe) [reply src/amp-story-embed.js]: I'm planning on using `parseUrlWithA`, which is what `parseUrlDeprecated` uses internally https://github.com/ampproject/amphtml/blob/2ca921f9e29fb54a872a0400d781fac15bb362bf/src/url.js#L119

Comment (jridgewell) [reply src/amp-story-embed.js]: If this is a non-amp page, then `parseUrlWithA` is appropriate.

**Comment by Enriqe** [comment]
Decided to move the loading one iframe + prerendering the next ones logic to another PR, to keep this one small and focused.

PTAL.

File: src/amp-story-embed.js [hunk 9]
```diff
@@ -150,7 +237,35 @@ self.onload = () => {
     const embed = embeds[i];
     const embedImpl = new AmpStoryEmbed(self, embed);
     embedImpl.buildCallback();
-    // TODO(Enriqe): add intersection observer that triggers layoutCallback().
-    embedImpl.layoutCallback();
+
+    if (IntersectionObserver && self === self.parent) {
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: Nit: please split the code into different functions/methods, the deepest code line has 7 levels of indentation, making it super hard to follow

**Comment by Enriqe** [comment]
PTAL @gmajoulet 

also need owners approval from @jridgewell / @choumx for `build-system/*` and `package.json`

File: src/amp-story-embed.js [hunk 10]
```diff
@@ -88,33 +125,67 @@ export class AmpStoryEmbed {
...
     setStyle(
-      this.iframeEl_,
+      iframeEl,
       'backgroundImage',
       story.getAttribute('data-poster-portrait-src')
     );
-    this.iframeEl_.classList.add('story');
-    this.initializeLoadingListeners_();
-    this.rootEl_.appendChild(this.iframeEl_);
+    iframeEl.setAttribute(IFRAME_ID_ATTR_NAME, 'i' + this.iframeIdCounter_++);
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: What's the plan for this in the future? It doesn't seem to be used at all in this PR
If we do need to keep it, could it be a className?

File: src/amp-story-embed.js [hunk 11]
```diff
@@ -88,33 +125,67 @@ export class AmpStoryEmbed {
...
+        console /*OK*/
+          .log({err});
+      }
+    );
+  }
+
+  /**
+   * @param {!Element} story
+   * @param {!Element} iframeEl
+   * @return {Promise<Messaging>}
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: Nit: `!Promise<!Messaging>`

File: src/amp-story-embed.js [hunk 12]
```diff
@@ -88,33 +125,67 @@ export class AmpStoryEmbed {
...
+   * @param {!Element} iframeEl
+   * @private
+   */
+  initializeLoadingListeners_(iframeEl) {
     this.rootEl_.classList.add(LoadStateClass.LOADING);
 
-    this.iframeEl_.onload = () => {
+    iframeEl.onload = () => {
       this.rootEl_.classList.remove(LoadStateClass.LOADING);
       this.element_.classList.remove(LoadStateClass.LOADING);
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: Not your doing, but this class doesn't seem to be ever added. Can we remove these statements?

File: src/amp-story-embed.js [hunk 13]
```diff
@@ -123,34 +194,145 @@ export class AmpStoryEmbed {
...
    * @public
    */
   layoutCallback() {
-    // TODO: Layout all child iframes.
-    return this.layoutIframe_(0);
+    if (this.isLaidOut_) {
+      return;
+    }
+
+    const iframes = toArray(this.rootEl_.querySelectorAll('iframe'));
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: As you create/recycle iframes as a user progresses through the stories, you won't be able to do `querySelectorAll` every time you want an iframe.
You should keep references to the iframes like you keep references to the `Messaging`. This will potentially change your data structure choice (`messagingFor`), so I believe this should be done in this PR.

File: src/amp-story-embed.js [hunk 14]
```diff
@@ -123,34 +194,145 @@ export class AmpStoryEmbed {
...
+    this.messagingFor_[iframeId].sendRequest(
+      'visibilitychange',
+      {state: 'visible'},
+      true
+    );
   }
 }
 
-self.onload = () => {
+/**
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: Should have mentioned that in my previous review, but this file will grow a lot, can you start thinking of an architecture to avoid having one single 3,000+ lines file?
All this build/layout logic looks like a good candidate to have its own file. Should it be a class or different helper functions? That's up to you, but I think the rest of the codebase uses almost exclusively classes, except for reusable helpers that have to be imported separately.

File: src/amp-story-embed.js [hunk 15]
```diff
@@ -123,34 +194,145 @@ export class AmpStoryEmbed {
...
+    iframe.setAttribute('src', href);
+  }
+
+  /**
+   * Gets encoded url for viewer usage.
+   * @param {string} href
+   * @return {!Location}
+   * @private
+   */
+  getEncodedUrl_(href) {
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: Please update the name of this method or the returned type, `URL` and `Location` are different objects

File: src/amp-story-embed.js [hunk 16]
```diff
@@ -123,34 +194,145 @@ export class AmpStoryEmbed {
...
+    const url = parseUrlWithA(this.cachedA_, location.href);
+
+    const params = dict({
+      'amp_js_v': '0.1',
+      'visibilityState': 'inactive',
+      'origin': url.origin,
+    });
+
+    let inputUrl = addParamsToUrl(href, params);
+    inputUrl = inputUrl.replace('amp_js_v=0.1', 'amp_js_v=0.1#');
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: Isn't this going to leave an extra `&` after the `#`?

Also, this won't work as `addParamsToUrl` seems to be adding the query parameter before the `hash`. You'd break the canonical URL by having the original hash at the very end of the URL.

File: test/unit/test-amp-story-embed.js [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+  let url;
+
+  function buildStoryEmbed() {
+    embedEl = win.document.createElement('amp-story-embed');
+    const storyAnchor = win.document.createElement('a');
+    url =
+      'https://www-washingtonpost-com.cdn.ampproject.org/v/s/www.washingtonpost.com/graphics/2019/lifestyle/travel/amp-stories/a-locals-guide-to-what-to-eat-and-do-in-new-york-city/';
+    storyAnchor.setAttribute('href', url);
+    embedEl.appendChild(storyAnchor);
+    storyEmbed = new AmpStoryEmbed(win, embedEl);
```
Comment (gmajoulet) [inline test/unit/test-amp-story-embed.js]: We should use the code you wrote to build the `AmpStoryEmbed`, to get more coverage.

File: test/unit/test-amp-story-embed.js [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+    win.document.body.appendChild(embedEl);
+  }
+
+  beforeEach(() => {
+    win = env.win;
+    buildStoryEmbed();
+  });
+
+  it('should build an iframe for each story', () => {
+    storyEmbed.buildCallback();
```
Comment (gmajoulet) [inline test/unit/test-amp-story-embed.js]: `buildCallback` and `layoutEmbed` should be called by `loadEmbeds`, not manually on each test

File: src/amp-story-embed.js (hunk 16 above)
Comment (Enriqe) [reply src/amp-story-embed.js]: Good catch! I am now extracting the original fragment param first before prepending it again after adding the new parameters.

File: src/amp-story-embed.js (hunk 10 above)
Comment (Enriqe) [reply src/amp-story-embed.js]: I had it to keep track and reference the iframes using their id's, which is why I wanted to use id or an attribute. But now that I am saving the reference to the iframes this is no longer needed.

**Comment by Enriqe** [comment]
PTAL @gmajoulet

**Comment by Enriqe** [comment]
Friendly ping

File: src/amp-story-embed-manager.js [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+  /**
+   * @param {!Window} win
+   * @constructor
+   */
+  constructor(win) {
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {?AmpStoryEmbed} */
+    this.currentEmbed_ = null;
```
Comment (gmajoulet) [inline src/amp-story-embed-manager.js]: Nit: we should not write code that's only used for unit testing purposes (this and the `getEmbed()` method). Do you have another way of writing your test that wouldn't need this piece of code?
`getEmbed` returns you a pretty arbitrary embed (the last one), and `currentEmbed` doesn't make sense if there are two embeds side to side in the page

File: src/amp-story-embed-manager.js [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+      };
+
+      const observer = new IntersectionObserver(intersectingCallback, {
+        rootMargin: '100%',
+      });
+      observer.observe(embedImpl.getElement());
+      return;
+    }
+
+    this.layoutFallback_(embedImpl);
```
Comment (gmajoulet) [inline src/amp-story-embed-manager.js]: Might be easier to read with an early return after this one line?

File: src/amp-story-embed-manager.js [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+  /**
+   * Fallback for when IntersectionObserver is not supported. Calls
+   * layoutCallback on the embed when it is close to the viewport.
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutFallback_(embedImpl) {
+    let tick = true;
+
+    this.win_.addEventListener('scroll', () => {
```
Comment (gmajoulet) [inline src/amp-story-embed-manager.js]: Is there an easy way to stop listening for scroll events once you instantiated all the embeds? If you think you might need this piece of code later for other reasons (play/pause embeds on scroll), feel free to simply add a comment.

File: src/amp-story-embed.js [hunk 22]
```diff
@@ -47,20 +57,48 @@ export class AmpStoryEmbed {
...
     );
 
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {!Object<number, Messaging>} */
+    this.messagingFor_ = {};
+
+    /** @private {!Array<!Element>} */
+    this.iframes_ = [];
```
Comment (gmajoulet) [inline src/amp-story-embed.js]: A few thoughts for another PR: I'm almost wondering if we should just store the story URLs instead of the iframes, so you can rotate through 3 or 5 iframe elements. Storing the actual iframes will make it harder/impossible to re-use the iframe elements.

File: src/amp-story-embed-manager.js (hunk 19 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Done.

File: src/amp-story-embed-manager.js (hunk 20 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Done.

File: src/amp-story-embed-manager.js (hunk 21 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Good point, I added a comment.

File: src/amp-story-embed.js (hunk 22 above)
Comment (Enriqe) [reply src/amp-story-embed.js]: Yep, this was one of the approaches I was thinking for my next PRs when enabling multiple stories and I agree we should rotate through a fixed amount of iframes.

File: src/amp-story-embed-manager.js [hunk 23]
```diff
@@ -0,0 +1,113 @@
...
+    this.win_ = win;
+  }
+
+  /**
+   * Calls layoutCallback on the embed when it is close to the viewport.
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutEmbed_(embedImpl) {
+    if (!IntersectionObserver || this.win_ !== this.win_.parent) {
```
Comment (jridgewell) [inline src/amp-story-embed-manager.js]: This will throw if `IntersectionObserver` is not defined. You want `this.win_.IntersectionObserver`, which won't.

File: src/amp-story-embed-manager.js [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+      this.layoutFallback_(embedImpl);
+      return;
+    }
+
+    const intersectingCallback = entries => {
+      entries.forEach(entry => {
+        if (!entry.isIntersecting) {
+          return;
+        }
+        embedImpl.layoutCallback();
```
Comment (jridgewell) [inline src/amp-story-embed-manager.js]: Why are you calling `layoutCallback`?

File: src/amp-story-embed-manager.js [hunk 25]
```diff
@@ -0,0 +1,113 @@
...
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutEmbed_(embedImpl) {
+    if (!IntersectionObserver || this.win_ !== this.win_.parent) {
+      this.layoutFallback_(embedImpl);
+      return;
+    }
+
+    const intersectingCallback = entries => {
```
Comment (jridgewell) [inline src/amp-story-embed-manager.js]: You should stop listening afterwards.

File: src/amp-story-embed-manager.js [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+      entries.forEach(entry => {
+        if (!entry.isIntersecting) {
+          return;
+        }
+        embedImpl.layoutCallback();
+      });
+    };
+
+    const observer = new IntersectionObserver(intersectingCallback, {
+      rootMargin: '100%',
```
Comment (jridgewell) [inline src/amp-story-embed-manager.js]: `rootMargin` will not work inside iframes. /cc @choumx

File: src/amp-story-embed-manager.js [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+   * Fallback for when IntersectionObserver is not supported. Calls
+   * layoutCallback on the embed when it is close to the viewport.
+   * @param {!AmpStoryEmbed} embedImpl
+   * @private
+   */
+  layoutFallback_(embedImpl) {
+    let tick = true;
+
+    // TODO(Enriqe): pause embeds when scrolling away from viewport.
+    this.win_.addEventListener('scroll', () => {
```
Comment (jridgewell) [inline src/amp-story-embed-manager.js]: You need to unregister this eventually.

File: src/amp-story-embed-manager.js (hunk 24 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: We want to lay out the component when it is close to the viewport, so in this case when the IO triggers.

File: src/amp-story-embed-manager.js (hunk 26 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Yep, I wrote a fallback when IO is not supported or if it's inside an iframe. See line 38.

File: src/amp-story-embed-manager.js (hunk 27 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: We will need it later to pause the component when the user scrolls away from the component.

File: src/amp-story-embed-manager.js (hunk 25 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Same as other comment, we will need it later to pause the stories when the user scrolls away from the embed.

File: src/amp-story-embed-manager.js (hunk 23 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Good to know, thanks!

File: src/amp-story-embed-manager.js (hunk 24 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Just to clarify and make sure you have the context: this is a pure JS component and is not built on AMP, which means it does not extend `AMP.BaseElement`.

**Comment by Enriqe** [comment]
PTAL @jridgewell @choumx

**Comment by Enriqe** [comment]
Ping @jridgewell

File: src/amp-story-embed-manager.js (hunk 27 above)
Comment (jridgewell) [reply src/amp-story-embed-manager.js]: You're not using it to pause now. Is this being added in a later PR?

File: src/amp-story-embed-manager.js [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+      if (!tick) {
```
Comment (jridgewell) [inline src/amp-story-embed-manager.js]: Nit: We have a `throttle` in `src/utils/rate-limit.js` that does exactly this.

File: src/amp-story-embed-manager.js (hunk 27 above)
Comment (Enriqe) [reply src/amp-story-embed-manager.js]: Correct, this will come in a follow up PR. 👍

**Comment by Enriqe** [comment]
Thanks all, merging!
